src/backend/parser/parse_utilcmd.c

```diff
@@ -3413,7 +3413,7 @@ transformIndexConstraint(Constraint *constraint, CreateStmtContext *cxt)
 		 * else dump and reload will produce a different index (breaking
 		 * pg_upgrade in particular).
 		 */
-		if (index_rel->rd_rel->relam != get_index_am_oid(DEFAULT_INDEX_TYPE, false))
+		if (index_rel->rd_rel->relam != get_index_am_oid(default_index_access_method, false))
```

Comment:
Suppose this case:

create table test(c1 int);
create unique index on test using btree(c1);
set default_index_access_method = 'brin';
alter table test add CONSTRAINT abc2 unique using index test_c1_idx;

The 'add constraint' statement fails here, and reports the ERROR 'index "test_c1_idx" is not a btree'，but index 'test_c1_idx' is indeed a btree.

Reply:
Its a bug and change to !IsIndexAccessMethod(BTREE_AM_OID, index_rel->rd_rel->relam)